send-transaction-service/src/transaction_client.rs

```diff
@@ -176,3 +213,237 @@ where
+{
+    runtime_handle: Handle,
+    sender: mpsc::Sender<TransactionBatch>,
+    // This handle is needed to implement `NotifyKeyUpdate` trait. It's only
+    // method takes &self and thus we need to wrap with Mutex.
+    join_and_cancel: Arc<Mutex<(Option<TpuClientJoinHandle>, CancellationToken)>>,
+    leader_updater: SendTransactionServiceLeaderUpdater<T>,
+    leader_forward_count: u64,
+}
+
+type TpuClientJoinHandle =
+    TokioJoinHandle<Result<TransactionStatsAndReceiver, ConnectionWorkersSchedulerError>>;
+
+impl<T> TpuClientNextClient<T>
+where
+    T: TpuInfoWithSendStatic + Clone,
+{
+    pub fn new(
+        runtime_handle: Handle,
+        my_tpu_address: SocketAddr,
+        tpu_peers: Option<Vec<SocketAddr>>,
+        leader_info: Option<T>,
+        leader_forward_count: u64,
+        identity: Option<Keypair>,
+    ) -> Self
+    where
+        T: TpuInfoWithSendStatic + Clone,
+    {
+        // The channel size represents 8s worth of transactions at a rate of
+        // 1000 tps, assuming batch size is 64.
+        let (sender, receiver) = mpsc::channel(128);
+
+        let cancel = CancellationToken::new();
+
+        let leader_info_provider = CurrentLeaderInfo::new(leader_info);
+        let leader_updater: SendTransactionServiceLeaderUpdater<T> =
+            SendTransactionServiceLeaderUpdater {
+                leader_info_provider,
+                my_tpu_address,
+                tpu_peers,
+            };
+        let config = Self::create_config(identity, leader_forward_count as usize);
+        let handle = runtime_handle.spawn(ConnectionWorkersScheduler::run(
+            config,
+            Box::new(leader_updater.clone()),
+            receiver,
+            cancel.clone(),
+        ));
+
+        Self {
+            runtime_handle,
+            join_and_cancel: Arc::new(Mutex::new((Some(handle), cancel))),
+            sender,
+            leader_updater,
+            leader_forward_count,
+        }
+    }
+
+    fn create_config(
+        stake_identity: Option<Keypair>,
+        leader_forward_count: usize,
+    ) -> ConnectionWorkersSchedulerConfig {
+        ConnectionWorkersSchedulerConfig {
+            bind: SocketAddr::new(Ipv4Addr::new(0, 0, 0, 0).into(), 0),
+            stake_identity,
+            // to match MAX_CONNECTIONS from ConnectionCache
+            num_connections: 1024,
+            skip_check_transaction_age: true,
+            // experimentally found parameter values
+            worker_channel_size: 64,
+            max_reconnect_attempts: 4,
+            leaders_fanout: Fanout {
+                connect: leader_forward_count,
+                send: leader_forward_count,
+            },
+        }
+    }
+
+    #[cfg(any(test, feature = "dev-context-only-utils"))]
+    pub fn cancel(&self) -> Result<(), Box<dyn std::error::Error>> {
+        let Ok(lock) = self.join_and_cancel.lock() else {
+            return Err("Failed to stop scheduler: TpuClientNext task panicked.".into());
```

Comment:
Probably don't need to have the bit about `TpuClientNext` task panicking. Presumably it is reporting its own issues

Reply:
fixed